vm/devices/storage/disk_get_vmgs/src/lib.rs

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! A disk backend using the GET's VMGS block interface.
+//!
+//! This is coded as a general-purpose block device (I guess you could boot a VM
+//! off of it), but it is likely only useful for using as the VMGS backing
+//! device.
+
+#![cfg(target_os = "linux")]
+#![warn(missing_docs)]
+
+use disk_backend::AsyncDisk;
+use disk_backend::DiskError;
+use disk_backend::SimpleDisk;
+use disk_backend::ASYNC_DISK_STACK_SIZE;
+use guest_emulation_transport::GuestEmulationTransportClient;
+use guestmem::MemoryRead;
+use guestmem::MemoryWrite;
+use inspect::Inspect;
+use save_restore::SavedBlockStorageMetadata;
+use scsi_buffers::RequestBuffers;
+use stackfuture::StackFuture;
+use std::io;
+use thiserror::Error;
+
+/// An implementation of [`SimpleDisk`] backed by the GET.
+#[derive(Clone, Debug, Inspect)]
+pub struct GetVmgsDisk {
+    get: GuestEmulationTransportClient,
+    sector_size: u32,
+    sector_shift: u32,
+    physical_sector_size: u32,
+    sector_count: u64,
+    max_transfer_sectors: u32,
+    max_transfer_size_bytes: u32,
+}
+
+/// An error that can occur when creating a new [`GetVmgsDisk`].
+#[derive(Debug, Error)]
+pub enum NewGetVmgsDiskError {
+    /// An IO error occurred while fetching the device info.
+    #[error("GET VMGS IO error")]
+    Io(#[source] guest_emulation_transport::error::VmgsIoError),
+    /// The sector size is not a power of two.
+    #[error("invalid sector size")]
+    InvalidSectorSize,
+    /// The physical sector size is not a power of two or is smaller than the sector size.
+    #[error("invalid physical sector size")]
+    InvalidPhysicalSectorSize,
+    /// The sector count is too large.
+    #[error("invalid sector count")]
+    InvalidSectorCount,
+    /// The disk ends with a partial physical sector.
+    #[error("disk ends with a partial physical sector")]
+    IncompletePhysicalSector,
+    /// The maximum transfer size is smaller than the physical sector size.
+    #[error("transfer size is smaller than the physical sector size")]
+    InvalidMaxTransferSize,
+}
+
+impl GetVmgsDisk {
+    /// Returns a new disk instance, communicating read and write IOs over the
+    /// `get` transport.
+    pub async fn new(get: GuestEmulationTransportClient) -> Result<Self, NewGetVmgsDiskError> {
+        let response = get
+            .vmgs_get_device_info()
+            .await
+            .map_err(NewGetVmgsDiskError::Io)?;
+        Self::new_inner(
+            get,
+            response.bytes_per_logical_sector.into(),
+            response.bytes_per_physical_sector.into(),
+            response.capacity,
+            response.maximum_transfer_size_bytes,
+        )
+    }
+
+    /// Create a disk using metadata previously-fetched via [`Self::save_meta`].
+    ///
+    /// # Caution
+    ///
+    /// `new_with_meta` does not confirm that the provided `meta` is what would
```

Comment:
stale comment (method name changed).

also, you sure you don't want to use the `dangerous_` prefix here?

Reply:
I don't think it really warrants `dangerous_`, any more than any of our other restore methods.